components/sup/src/manager/mod.rs

```diff
@@ -949,138 +1108,224 @@ impl Manager {
-        }
-
-        if self.user_config_watcher.remove(service).is_err() {
-            debug!(
-                "Error stopping user-config watcher thread for service {}",
-                service
-            );
-        }
-
-        self.updater.remove(service);
-    }
-
     /// Check if any elections need restarting.
     fn restart_elections(&mut self) {
         self.butterfly.restart_elections();
     }
 
-    fn shutdown(&mut self, cause: ShutdownReason) {
-        match cause {
-            ShutdownReason::PkgUpdating | ShutdownReason::Signal => {
-                // Previously, we would unconditionally set our health
-                // to departed. However, given our current conflation
-                // of Supervisor reachability and Service health, I
-                // suspect this causes instability when simply
-                // shutting down the Supervisor for a Supervisor
-                // upgrade.
-                //
-                // In the future, we may want to notify the rest of
-                // the network more formally of our intent to update,
-                // but for now, NOT convincing everyone that our
-                // services are gone seems like a good intermediate
-                // step. (We should also, of course, stop conflating
-                // Supervisor reachability and Service health!) Our
-                // existing suspicion mechanism should serve us well
-                // here if it takes longer than expected for the new
-                // Supervisor to come back up.
-                //
-                // Thus, for these given ShutdownReasons, we just
-                // won't send any Membership rumors out.
-            }
-            ShutdownReason::SvcStopCmd => {
-                // Just to call it out specifically, we shouldn't ever
-                // be called with this ShutdownReason.
-                //
-                // This is all being refactored elsewhere right now,
-                // for what it's worth.
-            }
-            ShutdownReason::LauncherStopping | ShutdownReason::Departed => {
-                // On the other hand, if we legitimately are going
-                // away, tell people, even though sending something
-                // out when _we've already been manually departed_ is
-                // perhaps excessive.
-                outputln!("Gracefully departing from butterfly network.");
-                self.butterfly.set_departed();
+    /// Create a future for stopping a Service. The Service is assumed
+    /// to have been removed from the internal list of active services
+    /// already.
+    fn stop(&self, service: Service) -> impl Future<Item = (), Error = ()> {
+        Self::service_stop_future(
+            service,
+            Arc::clone(&self.user_config_watcher),
+            Arc::clone(&self.updater),
+            Arc::clone(&self.busy_services),
+            Arc::clone(&self.reconcile_services),
```

Comment:
If I'm not mistaken, `Arc::clone()` does the same thing as just calling `.clone()` on an instance of an `Arc`, e.g. `&self.reconcile_services.clone()`.  I don't feel strongly one way or the other, but one advantage of just calling `.clone()` is it reads a bit easier.

Reply:
Yeah, it does do the same thing. I was mainly doing `Arc::clone` to drive home the fact that these clones are cheap, and that they're dealing with parts of the code that are part of our "giant ball of shared state". As we refactor and hopefully get to a place where we're not using so much shared state, this will help alert us to places we need to change.

Calling `.clone()` will work on anything that implements `Clone`, and not just `Arc`s. That means if, say, `updater` gets refactored away from an `Arc`, we could suddenly be cloning a much bigger thing without necessarily realizing that. Using `Arc::clone` will make that impossible.